plugins/follow_me/follow_me_impl.cpp

```diff
@@ -0,0 +1,390 @@
+    // needed by http://mavlink.org/messages/common#FOLLOW_TARGET
+    uint64_t elapsed_msec = static_cast<uint64_t>(_time.elapsed_since_s(now) * 1000); // milliseconds
+
+    _mutex.lock();
+//    LogDebug() << "Lat: " << _curr_target_location.latitude_deg << " Lon: " << _curr_target_location.longitude_deg <<
+//	" Alt: " << _curr_target_location.absolute_altitude_m;
+    const int32_t lat_int = static_cast<int32_t>(_curr_target_location.latitude_deg * 1e7);
+    const int32_t lon_int = static_cast<int32_t>(_curr_target_location.longitude_deg * 1e7);
+    const float alt = static_cast<float>(_curr_target_location.absolute_altitude_m);
+    _mutex.unlock();
+
+    const float pos_std_dev[] = { NAN, NAN, NAN };
+    const float vel[] = { NAN, NAN, NAN };
+    const float accel_unknown[] = { NAN, NAN, NAN };
+    const float attitude_q_unknown[] = { 1.f, NAN, NAN, NAN };
+    const float rates_unknown[] = { NAN, NAN, NAN };
+    uint64_t custom_state = 0;
+
+    mavlink_message_t msg {};
+    mavlink_msg_follow_target_pack(_parent->get_own_system_id(),
+                                   _parent->get_own_component_id(),
+                                   &msg,
+                                   elapsed_msec,
+                                   _estimatation_capabilities,
+                                   lat_int,
+                                   lon_int,
+                                   alt,
+                                   vel,
+                                   accel_unknown,
+                                   attitude_q_unknown,
+                                   rates_unknown,
+                                   pos_std_dev,
+                                   custom_state);
+
+    if (!_parent->send_message(msg)) {
+        LogErr() << "send_curr_target_location() failed..";
+    } else {
+        std::lock_guard<std::mutex> lock(_mutex);
+        _last_location = _curr_target_location;
+    }
+}
+
+void FollowMeImpl::stop_sending_target_location()
+{
+    // We assume that mutex was acquired by the caller
+    if (_curr_target_location_cookie) {
+        _parent->remove_call_every(_curr_target_location_cookie);
+        _curr_target_location_cookie = nullptr;
+    }
+    _mode = Mode::NOT_ACTIVE;
+}
+
+void FollowMeImpl::process_heartbeat(const mavlink_message_t &message)
+{
+    mavlink_heartbeat_t heartbeat;
+    mavlink_msg_heartbeat_decode(&message, &heartbeat);
+
+    bool follow_me_active = false; // tells whether we're in FollowMe mode right now
+    if (heartbeat.base_mode & MAV_MODE_FLAG_CUSTOM_MODE_ENABLED) {
+
+        px4::px4_custom_mode px4_custom_mode;
+        px4_custom_mode.data = heartbeat.custom_mode;
+
+        if (px4_custom_mode.main_mode == px4::PX4_CUSTOM_MAIN_MODE_AUTO &&
+            px4_custom_mode.sub_mode == px4::PX4_CUSTOM_SUB_MODE_AUTO_FOLLOW_TARGET) {
+            follow_me_active = true; // we're in FollowMe mode
+        }
+    }
+
+    {
+        _mutex.lock();
```

Comment:
For the case where you exit early with a return you can also use the `std::lock_guard` which unlocks when it goes out of scope.